python/src/graphql.rs

```diff
@@ -1,87 +1,590 @@
+    "Running server object has already been used, please create another one from scratch";
+
+/// A Raphtory server handler that also enables querying the server
+#[pyclass(name = "RunningRaphtoryServer")]
+pub(crate) struct PyRunningRaphtoryServer(Option<ServerHandler>);
+
+struct ServerHandler {
+    join_handle: JoinHandle<IoResult<()>>,
+    sender: Sender<()>,
+    client: PyRaphtoryClient,
+}
+
+impl PyRunningRaphtoryServer {
+    fn new(join_handle: JoinHandle<IoResult<()>>, sender: Sender<()>, port: u16) -> Self {
+        let url = format!("http://localhost:{port}");
+        Self(Some(ServerHandler {
+            join_handle,
+            sender,
+            client: PyRaphtoryClient::new(url),
+        }))
+    }
+
+    fn apply_if_alive<O, F>(&self, function: F) -> PyResult<O>
+    where
+        F: FnOnce(&ServerHandler) -> PyResult<O>,
+    {
+        match &self.0 {
+            Some(handler) => function(handler),
+            None => Err(PyException::new_err(RUNNING_SERVER_CONSUMED_MSG)),
+        }
+    }
+}
+
+#[pymethods]
+impl PyRunningRaphtoryServer {
+    /// Stop the server.
+    pub(crate) fn stop(&self) -> PyResult<()> {
+        self.apply_if_alive(|handler| {
+            handler
+                .sender
+                .send(())
+                .expect("Failed when sending cancellation signal");
+            Ok(())
+        })
+    }
+
+    /// Wait until server completion.
+    pub(crate) fn wait(mut slf: PyRefMut<Self>) -> PyResult<()> {
+        wait_server(&mut slf.0)
+    }
+
+    /// Wait for the server to be online.
+    ///
+    /// Arguments:
+    ///   * `millis`: the minimum number of milliseconds to wait (default 5000).
+    fn wait_for_online(&self, millis: Option<u64>) -> PyResult<()> {
+        self.apply_if_alive(|handler| handler.client.wait_for_online(millis))
+    }
+
+    /// Make a graphQL query against the server.
+    ///
+    /// Arguments:
+    ///   * `query`: the query to make.
+    ///   * `variables`: a dict of variables present on the query and their values.
+    ///
+    /// Returns:
+    ///    The `data` field from the graphQL response.
+    fn query(
+        &self,
+        py: Python,
+        query: String,
+        variables: Option<HashMap<String, PyObject>>,
+    ) -> PyResult<HashMap<String, PyObject>> {
+        self.apply_if_alive(|handler| handler.client.query(py, query, variables))
+    }
+
+    /// Send a graph to the server.
+    ///
+    /// Arguments:
+    ///   * `name`: the name of the graph sent.
+    ///   * `graph`: the graph to send.
+    ///
+    /// Returns:
+    ///    The `data` field from the graphQL response after executing the mutation.
+    fn send_graph(
+        &self,
+        py: Python,
+        name: String,
+        graph: MaterializedGraph,
+    ) -> PyResult<HashMap<String, PyObject>> {
+        self.apply_if_alive(|handler| handler.client.send_graph(py, name, graph))
+    }
+
+    /// Set the server to load all the graphs from its path `path`.
```

Comment:
shall we compress this and the function below and have overwrite as an optinal argument that defaults to false?

Reply:
Done